### PR TITLE
Add context param to Notes\DataStore::get_notes_where_clauses() and associated methods.

### LINKED ISSUE #8385: Proposal: ways to avoid performance issues from woocommerce_notes_where_clauses filter
Issue #7968 raised a performance concern for woocommerce.com brought about by the expense of the  [Notes DataStore](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Notes/DataStore.php)'s `get_notes()` method, which was being frequently called from inside woocommerce-admin. The query generated by `get_notes()` was expensive because of the extra conditions added in the `woocommerce_notes_where_clauses` filter by [WCCom_WC_Admin_Vendor_Notes_API](https://github.com/Automattic/woocommerce.com/blob/trunk/plugins/wccom-plugins/wc-admin-vendor-notes/includes/class-wccom-wc-admin-vendor-notes-api.php).

To address the immediate concern from the woocommerce.com side `filter_admin_notes_by_vendor_id()` was made more discriminating and from the woocommerce-admin side the notes that were the worst offenders were rewritten to avoid use of `get_notes()`, but the possibility of a recurrence of this problem remains as long as third party developers can add things to the where clause of `get_notes()` that will be executed in other contexts.

P2 discussion for context: pcQJnU-ah-p2

## Proposal 1: lookup_notes()

_Make a new lookup_notes method that is explicitly internal and bypasses the filter._

This leaves `get_notes()` alone but creates another method, `lookup_notes()` that works the same way except it doesn't apply the `woocommerce_notes_where_clauses` filter. The idea is that going forward anything in woo core or woocommerce admin that needs to query notes by something more complicated than what's already available in other methods (e.g. `get_notes_with_name()`) would use `lookup_notes()` instead of `get_notes()`.

### Pros

* No change required for third party developers at all.
* Our internal code probably shouldn't be using `get_notes()` for anything other than display anyway. Its results are paged!

### Cons

* Having two similarly named methods with the same signature, even if they're well documented, may well be confusing.

## Proposal 2: get_notes() context argument

_Add an optional context argument to get_notes for scoping custom where clauses._

Add a `'context'` key to the default arguments array for `get_notes()` with a default value of `'global'` (we'll use a constant for this rather than hardcoding it). Include explanatory comments suggesting that it be used in third-party code to identify `get_notes()` calls that need custom where clause insertions and that any filter functions added to `'woocommerce_notes_where_clauses'` should check for the `'context'` argument before making changes.

### Pros

* No change required in core code. (Even adding the default value is more by way of example than functionality.)
* If a third-party extension creates a performance issue it can be addressed within that extension's code.

### Cons

* Reliant on adoption by extension developers.

## Action

I'm creating PRs for both of these proposals, linked to this issue. We can adopt either, both, or neither. Please use this issue for discussion of the merits of the proposals and the individual PRs for discussion of their specific implementation.

### PR DESCRIPTION
**Pull request by jacob-sewell**
Fixes #8385 

(See the issue for much more discussion.)

This simply adds a `$context` param (default value 'global') to the method signatures of  `Admin\Notes\DataStore::get_notes()`, `get_notes_count()`, and `get_notes_where_clauses()` so that it can be used by third party developers to avoid applying custom `woocommerce_note_where_clauses` changes when not needed.

### Detailed test instructions:

There should be no change in behavior, but you can use a custom plugin to add a filter function to `'woocommerce_note_where_clauses'` and verify that it receives 'global' as its third parameter. For extra points you can call `get_notes()` and/or `get_notes_count()` yourself with a different value for `$context` and verify that it comes through as expected. See the new unit tests for ideas.

### CODE REVIEW

**Review by louwie17 — CHANGES_REQUESTED** [review]
I originally thought this change was far from necessary, but after some digging and pulling a few examples (I couldn't find many), I think this might be beneficial along with the change in #8387.

I do a have suggestion in terms of usage, I think instead of adding the context as an arg, I think we should create an extra parameter for the `get_notes` function called context and default to `global`.
We do the same when passing it into the `woocommerce_note_where_clauses` filter, adding context as a third parameter.
I think this help will prevent confusion, and would also help increase adoption as it will be more explicit.
Here is an example of Wordpress making use of $context in one of the filters - https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/class-wp-plugins-list-table.php#L908-L912

**Review by joshuatf — COMMENTED** [review]
Just popping in here to say that this is a fantastic addition.  More discussion here - p91TBi-7kt-p2 as to why we'll need this.

One thing that may be worth discussing prior to adding this is how we derive and set context.  Using enums/integers for context would be a bit more ideal for better indexing and searching, but we'd need a helper to determine context and some filters for extensibility.

**Comment by jacob-sewell** [comment]
> Just popping in here to say that this is a fantastic addition. More discussion here - p91TBi-7kt-p2 as to why we'll need this.

The approach contained here is essentially just a shibboleth that lets third-party code that messes with note filters turn itself off when the call is not coming from their own code. I'm not sure whether it can easily be extended to cover this case. Perhaps with the addition of `lookup_notes()` we can consider adding a global filter that screens out notes that shouldn't be shown on mobile when the query is for display? Does the backend know what kind of frontend is making requests of it?

File: src/Notes/DataStore.php [hunk 1]
```diff
@@ -442,7 +449,7 @@ public function get_notes_where_clauses( $args = array() ) {
 		 * @param string $where_clauses The generated WHERE clause.
 		 * @param array  $args          The original arguments for the request.
```
Comment (louwie17) [inline src/Notes/DataStore.php:450]: An update for the filter param comment would be great as well :)

**Review by louwie17 — APPROVED** [review]
Pre-approving this, thanks for the changes @jacob-sewell, I think this is great!
There was just one missing param comment, the one for the filter, if you could add that prior to merging :)

**Comment by ilyasfoo** [comment]
Hey, @jacob-sewell! Since it's past The Merge, I don't think we should merge any more codes in this repo unless it's cherry-picked into WCA 3.3 branch. Otherwise, will it be okay for you to recreate all the changes to the core repo instead?

**Comment by mattsherman** [comment]
@jacob-sewell We really should move this PR to core and get this wrapped up... I'll put this back in our backlog to tackle in the upcoming cycle.

**Comment by jacob-sewell** [comment]
Closing as obsolete (post-merge equivalent PR in woocommerce/woocommerce [here](https://github.com/woocommerce/woocommerce/pull/32574)).

**Pull request closed without merging**